apps/prairielearn/python/question_phases.py

```diff
@@ -110,13 +117,16 @@ def process_element(
                 sys.path.insert(0, str(pathlib.Path(course_path) / "serverFilesCourse"))
             sys.path.insert(0, str(element_path))
 
-            mod = {}
-            with open(element_controller_path, encoding="utf-8") as inf:
-                # use compile to associate filename with code object, so the
-                # filename appears in the traceback if there is an error
-                # (https://stackoverflow.com/a/437857)
-                code = compile(inf.read(), element_controller_path, "exec")
-                exec(code, mod)
+            mod = mod_cache.get(element_controller_path, None)
+            if mod is None:
+                mod = {}
+                with open(element_controller_path, encoding="utf-8") as inf:
+                    # Use `compile` to associate filename with code object, so the
+                    # filename appears in the traceback if there is an error:
+                    # https://stackoverflow.com/a/437857
+                    code = compile(inf.read(), element_controller_path, "exec")
+                    exec(code, mod)
```

Comment:
These two lines don't need to be under the `with` statement, only the line where the file is actually read.

Reply:
Done in 8173670.